apis/pkg/v1alpha1/revision_types.go

```diff
@@ -0,0 +1,124 @@
+/*
+Copyright 2020 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	rbac "k8s.io/api/rbac/v1"
+
+	runtimev1alpha1 "github.com/crossplane/crossplane-runtime/apis/core/v1alpha1"
+)
+
+// PackageRevisionDesiredState is the desired state of the package revision.
+type PackageRevisionDesiredState string
+
+const (
+	// PackageRevisionActive is an active package revision.
+	PackageRevisionActive PackageRevisionDesiredState = "Active"
+
+	// PackageRevisionInactive is an inactive package revision.
+	PackageRevisionInactive PackageRevisionDesiredState = "Inactive"
+)
+
+// PackageRevisionSpec specifies the desired state of a PackageRevision.
+type PackageRevisionSpec struct {
+	// Reference to install Pod. PackageRevision reads logs of this Pod to
+	// create resources owned by the PackageRevision.
+	InstallPodRef *runtimev1alpha1.Reference `json:"installPodRef,omitempty"`
+
+	// DesiredState of the PackageRevision. Can be either Active or Inactive.
+	DesiredState PackageRevisionDesiredState `json:"desiredState"`
+
+	// Image with PackageRevision contents.
+	Image string `json:"image"`
+
+	// Revision number. Indicates when the revision will be garbage collected
+	// based on the parent's RevisionHistoryLimit.
+	Revision int64 `json:"revision"`
+
+	// DependsOn is the list of packages and CRDs that this package depends on.
+	DependsOn []Dependency `json:"dependsOn,omitempty"`
+
+	// ControllerOptions is the options applied to packaged controller deployment.
+	ControllerOptions *ControllerOptions `json:"controllerOptions,omitempty"`
+
+	// Permissions are additional permissions that should be added to a packaged
+	// controller's service account.
+	Permissions *PermissionsSpec `json:"permissionSpec,omitempty"`
```

Comment:
This is copied over from existing package types, and I think you are right that they should be `permissionRequests`. In practice, the RBAC controller would look at these and add them to the role it creates for the SA. Since they are part of the `Spec`, I would expect them to be populated by the `Package` controller, rather than the `PackageRevision` controller, but that involves parsing the package contents in the `Package` controller. How would you feel about this being moved to the status of `___Revision` like we did with Objects?

Reply:
Regardless of what controller sets the requests, they do feel a little more like the desired state of the system to me. I can see why we might put them in status but I lean slightly toward keeping them in spec.

Reply:
The issue with putting them in `spec` is that if both the `Package` and `PackageRevision` controllers are watching the `PackageRevision`, then the `PackageRevision` controller updating the `spec` will trigger the the `Package` controller, which will then trigger the `PackageRevision` controller.... and so on. We are currently using the `APIPatchingApplicator`, and we could just say "if certain fields match, don't do a patch", but I would prefer if we were able to make it such that the `PackageRevision` controller didn't touch the `spec` at all. I do see your point that it feels like these belong in the `spec`, but it kind of depends on how you look at the revision. If the `spec` is indicating how it should parse and install a package, then you could argue that these don't belong in the `spec` since they are the _result_ of parsing the package, rather than indicating _how_ to parse a package.

Reply:
> it kind of depends on how you look at the revision.

Yeah, on reflection I think your approach of putting these in status makes sense. If I think of "`spec.installPodRef` as being the desired state - or a pointer to the actual source of desired state (the install pod's logs) - then modelling all these as actual state in status makes sense.